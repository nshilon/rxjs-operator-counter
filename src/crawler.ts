import * as Lint from 'tslint'
import * as path from 'path'
import { EOL } from 'os'
import { white, blue, green } from 'kleur'

export function crawl({ tsConfigPath = '' } = {}) {
  if (!tsConfigPath) {
    tsConfigPath = 'tsconfig.json'
    console.log(
      blue(`No tsconfig provided with -p, using "${tsConfigPath}" as fallback`)
    )
  }

  console.log(white('One moment, crawling files to find operators üßê'))

  const options = {
    fix: false,
    rulesDirectory: path.join(__dirname, 'rules'),
  }
  const program = Lint.Linter.createProgram(tsConfigPath, './')
  const linter = new Lint.Linter(options, program)
  const rules = new Map<string, Partial<Lint.IOptions>>([
    [
      'operator-counter',
      {
        ruleName: 'operator-counter',
      },
    ],
  ])
  const lintConfiguration = {
    rules,
    jsRules: rules,
    rulesDirectory: [options.rulesDirectory],
    extends: [''],
  }

  const files = Lint.Linter.getFileNames(program)
  files.forEach(file => {
    const fileContents = program.getSourceFile(file).getFullText()
    linter.lint(file, fileContents, lintConfiguration)
  })

  const results = linter.getResult()
  const hits = results.failures
    .map(x => x.getFailure())
    .reduce<{ [operator: string]: number }>((counter, operator) => {
      counter[operator] = (counter[operator] || 0) + 1
      return counter
    }, {})

  console.log(green().underline(`Results: ${EOL}`))

  Object.entries(hits)
    .sort(
      ([key1, count1], [key2, count2]) =>
        count2 - count1 || key1.localeCompare(key2)
    )
    .forEach(([key, count]) => console.log(green().bold(`${key}: ${count}`)))

  return hits
}
