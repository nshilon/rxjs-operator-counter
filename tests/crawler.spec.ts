import { crawl } from '../src/crawler'

test(`test run`, () => {
  let log = ''
  console.log = jest.fn(message => (log += message))
  crawl({
    tsConfigPath: './tests/fixtures/tsconfig.json',
  })

  expect(log).toMatchSnapshot()
})
