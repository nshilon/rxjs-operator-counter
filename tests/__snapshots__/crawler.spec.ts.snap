// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test run 1`] = `
"[37mOne moment, crawling files to find operators üßê[39m[4m[32mResults: 
[39m[24m[1m[32mmap: 4[39m[22m[1m[32mmapTo: 3[39m[22m[1m[32mcatchError: 1[39m[22m[1m[32mfilter: 1[39m[22m[1m[32mofType: 1[39m[22m[1m[32mselect: 1[39m[22m[1m[32mswitchMap: 1[39m[22m"
`;
